adal/src/main/java/com/microsoft/aad/adal/AcquireTokenSilentHandler.java

```diff
@@ -128,7 +129,34 @@ AuthenticationResult acquireTokenWithRefreshToken(final String refreshToken)
                 Logger.i(TAG, "Refresh token is not returned or empty", "");
                 result.setRefreshToken(refreshToken);
             }
-        } catch (final IOException | AuthenticationException exc) {
+        } catch (final AuthenticationException exc) {
+            //if get 503,504,500 && outageMode is on, return the stale token
+            if (exc.getCode().equals(ADALError.SERVER_NOT_RESPONDING) && mOutageModeIsOn) {
+                Logger.i(TAG, "The server is not responding after the retry with error code: " + exc.getCode(), "");
+                final TokenCacheItem accessTokenItem = mTokenCacheAccessor.getRegularRefreshTokenCacheItem(mAuthRequest.getResource(),
+                        mAuthRequest.getClientId(), mAuthRequest.getUserFromRequest());
+                if (accessTokenItem.getAccessToken() != null
+                        && accessTokenItem.getExtendedExpiresOn() != null
+                        && !TokenCacheItem.isTokenExpired(accessTokenItem.getExtendedExpiresOn())) {
+                    //While returning the stale token, the ExtendedExpiresOn property
+                    //should be reflected in ExpiresOn to minimize impact on client side code.
+                    accessTokenItem.setExpiresOn(accessTokenItem.getExtendedExpiresOn());
+                    final AuthenticationResult retryResult =  AuthenticationResult.createResult(accessTokenItem);
```

Comment:
How about this, moving this whole part into tokencacheAccessor? Add another method to just ask tokencacheaccessor give you a stale AT. Also, don't set the expiresOn with extendedExpiresOn on the tokenCacheItem itself, set it on the returned AuthenticationResult.